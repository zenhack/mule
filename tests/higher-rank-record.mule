#XFAIL?
# This tests that type annotations allow higher-rank types even when extracted
# from a record.
let r = {y = fn x. x} in
let f : all a. a -> a = r.y in
{ l = f {}
, k = f (fn z. z)
}
