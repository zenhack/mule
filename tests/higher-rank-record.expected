
Desugared:

	(letrec (types ())
	 (values ((r (letrec (types ()) (values ((y (fn x x)))) (((.= y) {}) y)))))
	 (letrec (types ()) (values ((f (: ((. y) r) (all a (a -> a))))))
	  (letrec (types ()) (values ((l (f {})) (k (f (fn z z)))))
	   (((.= l) (((.= k) {}) k)) l))))

inferred type:

	(record (types ())
	 (values ((l (record (types ()) (values ()))) (k (all a (a -> a))))))

Runtime term:

	(LetRec
	 (((0
	    (LetRec
	     (((0 (Lam (0 () (Var 0)))))
	      (App
	       ((App
	         ((Lam
	           (0 ()
	            (Lam (1 () (Update (old (Var 1)) (label y) (field (Var 0)))))))
	          (Record ())))
	        (Var 0)))))))
	  (LetRec
	   (((2 (App ((GetField y) (Var 1)))))
	    (LetRec
	     (((3 (App ((Var 2) (Record ()))))
	       (3 (App ((Var 2) (Lam (0 () (Var 0)))))))
	      (App
	       ((App
	         ((Lam
	           (0 ()
	            (Lam (1 () (Update (old (Var 1)) (label l) (field (Var 0)))))))
	          (App
	           ((App
	             ((Lam
	               (0 ()
	                (Lam (1 () (Update (old (Var 1)) (label k) (field (Var 0)))))))
	              (Record ())))
	            (Var 1)))))
	        (Var 0)))))))))

Evaluated:

	(Record ((k (Lam (0 () (Var 0)))) (l (Record ()))))
